android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch

```diff
@@ -0,0 +1,1112 @@
++
++	struct usb_endpoint_descriptor *isoc_tx_ep;
++	struct usb_endpoint_descriptor *isoc_rx_ep;
++
++	unsigned int sco_num;
++	int alternate_setting_num;
++	uint8_t *playback_data;
++
++	struct capture_data_cb *capture_data;
++
++	struct snd_pcm_substream *playback_stream;
++	struct snd_pcm_substream *capture_stream;
++
++	unsigned int playback_hwptr_done;	 ///* processed byte position in the buffer */
++	unsigned int playback_transfer_done;	///* processed frames since last period update */
++	unsigned int capture_hwptr_done;
++	unsigned int capture_transfer_done;
++
++	struct usb_anchor rx_anchor;
++
++	struct urb *tx_urb;
++	struct urb *rx_urb[MAX_URBS];
++};
++
++static void btusb_isoc_tx_complete(struct urb *urb);
++
++uint8_t handle[2];
++
++static const struct usb_device_id btusb_sco_table[] = {
++	/* Intel Bluetooth devices */
++	{ USB_DEVICE(0x8087, 0x0025), .driver_info = BTUSB_INTEL_NEW },
++	{ USB_DEVICE(0x8087, 0x0026), .driver_info = BTUSB_INTEL_NEW },
++	{ USB_DEVICE(0x8087, 0x0a2b), .driver_info = BTUSB_INTEL_NEW },
++	{ USB_DEVICE(0x8087, 0x0aaa), .driver_info = BTUSB_INTEL_NEW },
++	{ USB_DEVICE(0x8087, 0x0aa7), .driver_info = BTUSB_INTEL_NEW },
++	{ }	/* Terminating entry */
++};
++
++MODULE_DEVICE_TABLE(usb, btusb_sco_table);
++
++static void dummy_proc_init(struct snd_card *card)
++{
++	struct snd_info_entry *entry;
++	if (!snd_card_proc_new(card, "dummy_pcm", &entry)) {
++		printk("btusb_sco:Registred sounc card in proc");
++	}
++}
++
++
++static inline void __fill_isoc_descriptor(struct urb *urb, int len, int mtu)
++{
++	int i, offset = 0;
++
++	for (i = 0; i < BTUSB_MAX_ISOC_FRAMES && len >= mtu;
++			i++, offset += mtu, len -= mtu) {
++		urb->iso_frame_desc[i].offset = offset;
++		urb->iso_frame_desc[i].length = mtu;
++	}
++
++	if (len && i < BTUSB_MAX_ISOC_FRAMES) {
++		urb->iso_frame_desc[i].offset = offset;
++		urb->iso_frame_desc[i].length = len;
++		i++;
++	}
++
++	urb->number_of_packets = i;
++}
++
++static int btusb_isoc_initialize_tx_urb(struct btusb_data *data)
++{
++	unsigned int pipe;
++
++	if (!data->isoc_tx_ep) {
++		pr_emerg("%s - isoc_tx_ep is NULL\n", __func__);
++		return -ENODEV;
++	}
++
++	data->tx_urb = usb_alloc_urb(BTUSB_MAX_ISOC_FRAMES, GFP_KERNEL);
++	if (!data->tx_urb) {
++		pr_emerg("%s - usb_alloc_urb is NULL\n", __func__);
++		return -ENOMEM;
++	}
++
++	pipe = usb_sndisocpipe(data->udev, data->isoc_tx_ep->bEndpointAddress);
++
++	usb_fill_int_urb(data->tx_urb, data->udev, pipe, data->playback_data,
++			MAX_PACKET_SIZE, btusb_isoc_tx_complete, data,
++			data->isoc_tx_ep->bInterval);
++
++	data->tx_urb->transfer_flags = URB_ISO_ASAP;
++
++	__fill_isoc_descriptor(data->tx_urb, MAX_PACKET_SIZE,
++			le16_to_cpu(data->isoc_tx_ep->wMaxPacketSize));
++
++	return 0;
++}
++
++static void btusb_isoc_prepare_tx_urb(struct btusb_data *data)
++{
++	unsigned int stride, frames, bytes;
++	static unsigned int oldptr;
++	uint8_t period_elapsed = 0;
++	unsigned long flags;
++	uint8_t *hw_ptr;
```

Comment:
hw_ptr name looks misleading here. Please rename it to something better.. may be sco_buf or